core/src/main/java/io/confluent/kafka/schemaregistry/rest/resources/SubjectVersionsResource.java

```diff
@@ -624,18 +629,46 @@ public void modifyTags(
     }
     schemaRegistry.getCompositeUpdateRequestHandler().handle(schema, request);
 
-    if (request.getRuleSet() != null) {
-      try {
-        request.getRuleSet().validate();
-      } catch (RuleException e) {
-        throw new RestInvalidRuleSetException(e.getMessage());
-      }
-    }
-
     Map<String, String> headerProperties = requestHeaderBuilder.buildRequestHeaders(
         headers, schemaRegistry.config().whitelistHeaders());
     RegisterSchemaResponse registerSchemaResponse;
     try {
+      RuleSet ruleSet = request.getRuleSet();
+      if (request.getRulesToMerge() != null || request.getRulesToRemove() != null) {
+        int oldVersion = request.getNewVersion() != null ? request.getNewVersion() - 1 : -1;
+        Schema oldSchema = schemaRegistry.get(subject, oldVersion, false);
+        // Use the previous ruleSet instead of the passed in one
+        ruleSet = oldSchema != null ? oldSchema.getRuleSet() : null;
+        if (request.getRulesToMerge() != null) {
+          ruleSet = mergeRuleSets(ruleSet, request.getRulesToMerge());
+        }
+        if (ruleSet != null && request.getRulesToRemove() != null) {
+          List<String> rulesToRemove = request.getRulesToRemove();
+          List<Rule> migrationRules = ruleSet.getMigrationRules();
+          if (migrationRules != null) {
+            migrationRules = migrationRules.stream()
+                .filter(r -> !rulesToRemove.contains(r.getName()))
+                .collect(Collectors.toList());
+          }
+          List<Rule> domainRules = ruleSet.getDomainRules();
+          if (domainRules != null) {
+            domainRules = domainRules.stream()
+                .filter(r -> !rulesToRemove.contains(r.getName()))
+                .collect(Collectors.toList());
+          }
+          ruleSet = new RuleSet(migrationRules, domainRules);
+        }
+        request.setRuleSet(ruleSet);
+      }
+
+      if (request.getRuleSet() != null) {
+        try {
+          request.getRuleSet().validate();
+        } catch (RuleException e) {
+          throw new RestInvalidRuleSetException(e.getMessage());
+        }
+      }
+
```

Comment:
Is it worth putting this code block in a function since its just modifying the `request`?